src/lib/reasoners/bitv.ml

```diff
@@ -1451,11 +1443,17 @@ module Shostak(X : ALIEN) = struct
   let solve r1 r2 pb =
     Sig.{pb with sbt = List.rev_append (solve_bis r1 r2) pb.sbt}
 
-  (* Pop the first bit, raises [Not_found] if there is no first bit *)
+  (* Pop the first bit, raises [Not_found] if there is no first bit.
+
+     Note that the returned bv has an incorrect size. *)
   let pop_bit = function
     | [] -> raise Not_found
-    | { bv = Cte b; sz } as st :: rst ->
-      Some b, if sz > 1 then { st with sz = sz - 1 } :: rst else rst
+    | ({ bv = Cte n; sz } as st) :: rst ->
+      Some (Z.testbit n (sz - 1)),
+      if sz > 1 then
+        { st with sz = sz - 1 } :: rst
+      else
+        rst
     | { bv = Other _ | Ext _ ; sz } as st :: rst ->
       None, if sz > 1 then { st with sz = sz - 1 } :: rst else rst
 
```

Comment:
Unrelated to the PR, but the comments in lines 1467-1668 should be fixed/reworded, with something like `_ are the values we must (differentiate|distinguish) from (those) that start* with a _`

Reply:
Can you clarify what should be fixed? The values in `[t]` all start with a `1`, and the model value must be distinct from all these values; the values in `[f]` all start with a `0` and the model value must also be distinct from all these values. I wouldn't understand that from your reformulation.

Reply:
It's just the wording, the current comments say:
```
    (* [t] are the values we must be distinct from that starts with a '1' *)
    (* [f] are the values we must be distinct from that starts with a '0' *)
```
It's true that what I suggested doesn't help. But I don't get your explanation from the current comments either. 
Putting the explanation you gave as a comment instead of the two comments that are currently present would be better IMO.

Reply:
Right. Would something like "[t] holds the values that (a) we must be distinct from and (b) start with a `1`" be clearer? I will make a PR once this one is merged to avoid conflicts.

Reply:
Yes, it would be clearer.